docker.py

```diff
@@ -193,22 +194,31 @@ def load(readable):
     return loaded_image
 
 
-def build_and_tag(tag_name, dockerfile_path, cwd=None, log_path=None):
-    print("Building runner Docker image %s from %s" % (tag_name,
-                                                       dockerfile_path))
+def build_and_tag(tag_name, dockerfile_path, cwd=None,
+                  log_path=None, args=None):
+    if args is None:
+        args = {}
+    build_args = list(reduce(lambda a, b: a + b,
+                             map(lambda k: [
+                                 "--build-arg",
+                                 "%s=%s" % (k, args[k])
+                             ], args), []))
+
     if not log_path:
         subprocess.check_call([
-            "docker", "build", "--tag", tag_name, dockerfile_path
+            "docker", "build", *build_args, "--tag", tag_name, dockerfile_path
         ], cwd=cwd)
     else:
         clean_tag = re.sub(r"\W", "_", tag_name)
         out_path = os.path.join(log_path, "build_{}_out.log".format(clean_tag))
         err_path = os.path.join(log_path, "build_{}_err.log".format(clean_tag))
         with open(out_path, "w") as out_fd:
             with open(err_path, "w") as err_fd:
-                print(["docker", "build", "--tag", tag_name, dockerfile_path])
+                print(["docker", "build", *build_args,
+                       "--tag", tag_name, dockerfile_path])
                 subprocess.check_call([
-                    "docker", "build", "--tag", tag_name, dockerfile_path
+                    "docker", "build", *build_args,
+                    "--tag", tag_name, dockerfile_path
```

Comment:
Optional: at this point, it might be a smart idea to assemble the list once, store it in a variable (e.g.,
`cmd`) and use it in all 3 places.

Reply:
Sure